cdi/src/test/java/brave/webmvc/TracingProducerTest.java

```diff
@@ -0,0 +1,153 @@
+import zipkin2.Span;
+import zipkin2.reporter.Reporter;
+
+@EnableAutoWeld
+@AddBeanClasses({ConfigProducer.class, TracingProducer.class})
+@EnableAlternatives(TracingProducerTest.class)
+class TracingProducerTest {
+
+    private static final Logger log = Logger.getLogger(TracingProducerTest.class.getName());
+
+    private static final TestSpanReporter REPORTER = new TestSpanReporter();
+    private static final MockWebServer SERVER = new MockWebServer();
+
+    public interface TestInterface {
+        @GET
+        @Path("test")
+        @Produces(MediaType.TEXT_PLAIN)
+        String test();
+    }
+
+    @BeforeAll
+    static void beforeAll() throws IOException {
+        SERVER.start();
+    }
+
+    @AfterAll
+    static void afterAll() throws IOException {
+        SERVER.shutdown();
+    }
+
+    @AfterEach
+    void beforeEach() {
+        REPORTER.clear();
+    }
+
+    @Test
+    void producesTracing() {
+        assertNotNull(TracingProducer.createTracing());
+    }
+
+    @Test
+    void tracesRequest() {
+        SERVER.enqueue(new MockResponse()
+                .setResponseCode(HttpServletResponse.SC_OK)
+                .addHeader("Content-Type", MediaType.TEXT_PLAIN)
+                .setBody("response-text"));
+
+        final String response = new BuilderResolver().newBuilder()
+                .baseUri(SERVER.url("").uri())
+                .register(TracingClientFilter.create(
+                        HttpTracing.create(TracingProducer.createTracing())
+                                .clientOf("test-service")))
+                .build(TestInterface.class)
+                .test();
+        assertEquals("response-text", response);
+        assertEquals(1, REPORTER.getSpans().size(), "Wrong number of reported spans");
+    }
+
+    /**
+     * As the tracing filter is not processed no span is reported.
+     * this is due to a wrong TracingClientFilter order / priority.
+     *
+     * @throws InterruptedException if response could not be taken
+     */
+    @Test
+    void noTraceOnErrorResponse() throws InterruptedException {
+        SERVER.enqueue(new MockResponse().setResponseCode(SC_INTERNAL_SERVER_ERROR));
+
+        try {
+            new BuilderResolver().newBuilder()
+                    .baseUri(SERVER.url("").uri())
+                    .register(TracingClientFilter.create(
+                            HttpTracing.create(TracingProducer.createTracing())
+                                    .clientOf("test-service")))
+                    .build(TestInterface.class)
+                    .test();
+        } catch (final WebApplicationException e) {
+            log.log(Level.WARNING, "Received expected error response: " + e.getMessage());
+        }
+
+        assertNotNull(SERVER.takeRequest(5, TimeUnit.SECONDS), "No recorded request to mock-server");
+        // FIXME should be 1 instead of 0!
+        assertEquals(0, REPORTER.getSpans().size(), "Wrong number of reported spans");
```

Comment:
you don't need to do unit tests, as it complicates the project and also makes distracting dependencies. The goal of these examples are to focus attention only on what is deserving of it. For example, that the Frontend and Backend controllers have zero tracing apis and everything is handled outside.

Let's save things like this for a brave-cdi project, as I would have a ton more feedback for you there, for example, that brave-tests includes a reporter rule you should use :P Before we start brave-cdi, let's just KISS here. when it works, move on to more. sg?